File: python/mandelbulb_singlethread/mandelbulb_singlethread.py
from PIL import Image
import numpy as np
import time

WIDTH = 2048
HEIGHT = 2048

# Maximum number of steps for ray marching
MAX_MARCHING_STEPS = 100
# Maximum marching distance before stopping
MAX_DISTANCE = 100.0
# Minimum distance threshold to consider on surface
SURFACE_DISTANCE = 0.001

# Maximum iterations for distance estimator function
MAX_ITERATIONS = 8
# Escape radius for distance estimator
BAILOUT = 2.0
# Power exponent for Mandelbulb fractal shape
POWER = 8.0

# Distance estimator (DE) function for the Mandelbulb fractal
# Calculate minimum distance from point to fractal
def mandelbulbDE(pos):
  # Position
  z = pos.copy()
  # Distance increment
  dr = 1.0
  # Distance
  r = 0.0

  for i in range(MAX_ITERATIONS):
    r = np.linalg.norm(z)
    if (r > BAILOUT): break

    # Convert to spherical coordinates
    theta = np.acos(z[2] / r)
    phi = np.atan2(z[1], z[0])
    zr = pow(r, POWER)

    # Scale the distance increment
    dr = pow(r, POWER - 1.0) * POWER * dr + 1.0

    # Calculate new coordinates
    sinTheta = np.sin(POWER * theta)
    cosTheta = np.cos(POWER * theta)
    sinPhi = np.sin(POWER * phi)
    cosPhi = np.cos(POWER * phi)

    z = np.array([zr * sinTheta * cosPhi, zr * sinTheta * sinPhi, zr * cosTheta])
    z = z + pos

  return 0.5 * np.log(r) * r / dr

# Calculate the normal vector at a surface point
def getNormal(pos):
  eps = 0.001
  dx = mandelbulbDE(pos + np.array([eps, 0, 0])) - mandelbulbDE(pos - np.array([eps, 0, 0]))
  dy = mandelbulbDE(pos + np.array([0, eps, 0])) - mandelbulbDE(pos - np.array([0, eps, 0]))
  dz = mandelbulbDE(pos + np.array([0, 0, eps])) - mandelbulbDE(pos - np.array([0, 0, eps]))
  normal = np.array([dx, dy, dz])
  return normal / np.linalg.norm(normal) # normalize

# Ray marching function to find distance to the surface
def rayMarch(f, dir):
  distance = 0.0
  for i in range(MAX_MARCHING_STEPS):
    pos = f + dir * distance
    d = mandelbulbDE(pos)
    if (d < SURFACE_DISTANCE):
      # Intersection
      return distance
    if (distance > MAX_DISTANCE):
      # No intersection
      break
    distance += d
  # No intersection found
  return -1.0

def mandelbulb():
  image = np.zeros(shape=(WIDTH, HEIGHT, 3), dtype=np.uint8)

  cameraPos = np.array([0.0, 0.0, -1.6])
  lightDir = np.array([1.0, 1.0, -1.0])
  lightDir = lightDir / np.linalg.norm(lightDir) # normalize

  for y in range(HEIGHT):
    for x in range(WIDTH):
      # Normalize pixel coordinates
      px = (x / WIDTH) * 2.0 - 1.0
      py = (y / HEIGHT) * 2.0 - 1.0
      # Adjust for aspect ratio
      px *= WIDTH / HEIGHT

      # Calculate ray direction
      dir = np.array([px, py, 1.0])
      dir = dir / np.linalg.norm(dir) # normalize

      dist = rayMarch(cameraPos, dir)
      if (dist > 0.0):
        # Calculate intersection point and normal
        pos = cameraPos + np.dot(dir, dist)
        normal = getNormal(pos)

        # Compute lighting (Lambertian shading)
        diffuse = np.max([0.0, np.dot(normal, lightDir)])

        # Set pixel color based on lighting
        color = diffuse * 255
        image[x][y] = [color, color, color]
      else:
        # Set background color
        image[x][y] = [0, 0, 0]

  return image

STEPS = 10

print("[PYTHON] Mandelbulb single thread benchmark")
print(f"Size: {WIDTH} x {HEIGHT}\nPOWER: {POWER} / MAX ITERATIONS: {MAX_MARCHING_STEPS}")

for i in range(STEPS):
  print(f"Start rendering {i + 1}/{STEPS}")

  start = time.time_ns()
  data = mandelbulb()
  duration = time.time_ns() - start

  print(f"[{i + 1}] Execution time: {duration} ns ({duration / 1000000000} s)")

  if (i == 0):
    print("Save first run to file mandelbulb.png")
    image = Image.fromarray(data, 'RGB')
    image.save('mandelbulb.png')

print("All renderings are done")

File: js/bin/mandelbulb_singlethread.js
#!/usr/bin/env node

import { measure, Vec3, mandelbulblib } from "../lib/index.js";
import fs from "fs";
import { createCanvas } from "canvas";

// Image width
const WIDTH = 2048;
// Image height
const HEIGHT = 2048;

// Number of renders to be performed (identical renders for quality check)
const STEPS = 10;

function mandelbulb() {
  const image = createCanvas(WIDTH, HEIGHT);
  const ctx = image.getContext("2d");
  const imageData = ctx.createImageData(WIDTH, HEIGHT);
  const buffer = imageData.data;

  const cameraPos = new Vec3(0.0, 0.0, -1.6);
  const lightDir = new Vec3(1.0, 1.0, -1.0).normalize();

  for (let y = 0; y < HEIGHT; y++) {
    for (let x = 0; x < WIDTH; x++) {
      // Normalize pixel coordinates
      const px = (x / WIDTH) * 2.0 - 1.0;
      const py = (y / HEIGHT) * 2.0 - 1.0;
      // Adjust px for aspect ratio
      const aspectRatio = WIDTH / HEIGHT;
      const adjustedPx = px * aspectRatio;

      // Calculate ray direction
      const dir = new Vec3(adjustedPx, py, 1.0).normalize();
      const dist = mandelbulblib.rayMarch(cameraPos, dir);
      const index = (x + y * WIDTH) * 4;
      if (dist > 0.0) {
        // Calculate intersection point and normal
        const pos = cameraPos.add(dir.multiply(dist));
        const normal = mandelbulblib.getNormal(pos);

        // Compute lighting (Lambertian shading)
        const diffuse = Math.max(0.0, normal.dot(lightDir));

        // Set pixel color based on lighting
        const color = Math.floor(diffuse * 255);
        buffer[index] = color;
        buffer[index + 1] = color;
        buffer[index + 2] = color;
        buffer[index + 3] = 255;
      } else {
        // Set background color
        buffer[index] = 0;
        buffer[index + 1] = 0;
        buffer[index + 2] = 0;
        buffer[index + 3] = 255;
      }
    }
  }

  // Save image data and return it
  ctx.putImageData(imageData, 0, 0);
  return image;
}

//////

(async () => {
  console.log("[JS] Mandelbulb single thread benchmark");
  console.log(
    `Size: ${WIDTH} x ${HEIGHT}\nPOWER: ${mandelbulblib.POWER} / MAX ITERATIONS: ${mandelbulblib.MAX_MARCHING_STEPS}`,
  );

  // Steps
  for (let i = 0; i < STEPS; i++) {
    console.log(`Start rendering ${i + 1}/${STEPS}`);
    const { result: image, duration } = await measure(mandelbulb);
    console.log(
      `[${i + 1}] Execution time: ${duration} ns (${Number(duration) / 1e9} s)`,
    );

    if (i === 0) {
      console.log("Save first run to file mandelbulb.png");
      const out = fs.createWriteStream("mandelbulb.png");
      const stream = image.createPNGStream();
      stream.pipe(out);
      await new Promise((resolve) => out.on("finish", resolve));
    }
  }
})();
